Tutorial update (#9940)
PrefectHQ/prefect: docs/tutorial/deployments.md

description: Learn how Prefect flow deployments enable configuring flows for scheduled and remote execution.
tags:
    - work pools
    - workers
    - orchestration
    - flow runs
    - deployments
    - schedules
    - triggers
    - tutorial
search:
  boost: 2
---
# Deploying Flows

## Why Deploy

> The most common reason to use a tool like Prefect, is [scheduling](/concepts/schedules). You want your flows running on production infrastructure in a consistent and predictable way. Up to this point, we’ve demonstrated running Prefect flows as scripts, but this means *you* have been the one triggering flow runs. In order to schedule flow runs or trigger them based on [events](/cloud/events/) you’ll need to understand Prefect’s concept of a flow's [Deployment](/concepts/deployments/).

Reviewer: ```suggestion
One of the most common reasons to use a tool like Prefect is [scheduling](/concepts/schedules). You want your flows running on production infrastructure in a consistent and predictable way. Up to this point, we’ve demonstrated running Prefect flows as scripts, but this means *you* have been the one triggering flow runs. In order to schedule flow runs or trigger them based on [events](/cloud/events/) you’ll need to [deploy](/concepts/deployments/) them.
```

A deployed flow gets the following additional cababilities:

- flows triggered by scheduling
- remote execution of flows triggered from the UI
- flow triggered by automations or events

## What is a Deployment

> Deploying your flows is, in essence, the act of informing the Prefect API of:

Reviewer: ```suggestion
Deploying your flows is, in essence, the act of informing Prefect of:
```

1. Where to run your flows
2. How to run your flows
3. When to run your flows

> This information is encapsulated and sent to Prefect as a [Deployment](/concepts/deployments/) which becomes a server side object containing the crucial metadata needed for orchestration. Deployments elevate workflows from functions that you call manually to API-managed entities.

Reviewer: ```suggestion
This information is encapsulated and sent to Prefect as a [Deployment](/concepts/deployments/) which contains the crucial metadata needed for orchestration. Deployments elevate workflows from functions that you call manually to API-managed entities.
```

Attributes of a deployment include (but are not limited to):

- __Flow entrypoint__: path to your flow function would start the flow

> - __Workpool__: points to the infrastructure you want your flow to run in

Reviewer: ```suggestion
- __Work pool__: points to the infrastructure you want your flow to run in
```

- __Schedule__: optional schedule for this deployment
- __Tags__: optional metadata

Before you build your first deployment, its helpful to understand how Prefect configures flow run infrastrucuture. This means setting up a work pool and a worker to enable **your flows** to run on **your infrastructure**.

## Why Workpools and Workers

> Easily running Prefect flows locally is great for testing and development purposes. But for production settings, Prefect provides you with the worker and work pool concepts to allow you to run flows in the environments best suited to their execution.

Reviewer: ```suggestion
Running Prefect flows locally is great for testing and development purposes. But for production settings, Prefect work pools and workers allow you to run flows in the environments best suited to their execution.
```

> Workers and work pools bridge the Prefect orchestration layer with the infrastructure the actual flows need to be executed on.

Reviewer: ```suggestion
Workers and work pools bridge the Prefect orchestration layer with the infrastructure the flows are actually executed on.
```

> You can configure work pools within the Prefect UI or server. They prioritize the flows and respond to polling from the worker. Workers are light-weight, pieces of infrastructure where flows get executed.

Reviewer: ```suggestion
You can configure work pools within the Prefect UI. They prioritize the flows and respond to polling from the worker. Workers are light-weight processes that run in the environment where flows are executed.
```

**Work Pools:**

- Organize the flows for your worker to pick up and execute.
- Describe the ephemeral infrastructure configuration that the worker will create for each flow run.

> - Prioritize the flows and respond to polling from its worker.

Reviewer: ```suggestion
- Prioritize the flows and respond to polling from its worker(s).
```

> - They prioritize the flows and respond to polling from its worker.

Reviewer: ```suggestion
- Prioritize the flows and respond to polling from its worker.
```

        worker["Worker"]
				subgraph flow_run_infra[Flow Run Infra]
					flow_run(("Flow Run"))
				end
        
    end

    subgraph api["Prefect API"]
				deployment --> |assigned to| work_pool
        work_pool(["Work Pool"])
    end

    worker --> |polls| work_pool
    worker --> |creates| flow_run_infra
```

!!! note "Security Note:"
    Prefect provides execution through the hybrid model which allows you to deploy workflows that run in the environments best suited to their execution while allowing you to keep your code and data completely private. There is no ingress required. For more information see [here.](https://www.prefect.io/security/overview/#overview)

Now that we’ve reviewed the concepts of a Work Pool and Worker, let’s create them so that you can deploy your tutorial flow, and execute it later using the Prefect Orchestration API.

## Setting up the Worker and Work Pool
For this tutorial you will create a *process type* work pool via the CLI. 

The process work pool type specifies that all work sent to this work pool will run as a subprocess inside the same infrastructure from which the worker is started.

!!! tip "Other Workpools"

>     Aside from process, there are a variety of different work pool types you might consider in a production setting to containerize your flow runs that leverage managed execution platforms, like Kubernetes services or serverless computing environments such as AWS ECS, Azure Container Instances, or GCP Cloud Run which are expanded upon in the guides section.

Reviewer: ```suggestion
    There are work pool types for all major managed code execution platforms, such as Kubernetes services or serverless computing environments such as AWS ECS, Azure Container Instances, or GCP Cloud Run, which are expanded upon in the guides section.
```

In your terminal run the following command to set up a work pool. 
<div class="terminal">
```bash
prefect work-pool create --type process my-process-pool
```
</div>

> Now that you have created the work pool, let’s confirm that the work pool was successfully created by running the following command in the same terminal.  You should see your new `my-process-pool` in the output list.

Reviewer: ```suggestion
Let’s confirm that the work pool was successfully created by running the following command in the same terminal. You should see your new `my-process-pool` in the output list.
```

<div class="terminal">
```bash
prefect work-pool ls 
```
</div>

> Finally, let’s double check that you can see this work pool in the Prefect Cloud UI . Navigate to the Work Pool tab and verify that you see `my-process-pool` listed.

Reviewer: ```suggestion
Finally, let’s double check that you can see this work pool in the Prefect Cloud UI. Navigate to the Work Pool tab and verify that you see `my-process-pool` listed.
```

> Finally, let’s double check in the Prefect Cloud UI that you can see this work pool. Navigate to the Work Pool tab and verify that you see `my-process-pool` listed.

Reviewer: ```suggestion
Finally, let’s double check that you can see this work pool in the Prefect Cloud UI . Navigate to the Work Pool tab and verify that you see `my-process-pool` listed.
```

<div class="terminal">
```bash
                                             Work Pools                                             
┏━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┓
┃ Name                  ┃ Type          ┃                                   ID ┃ Concurrency Limit ┃
┡━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━┩
│ my-process-pool       │ process       │ 33ce63a5-cc80-4f43-9092-11122ccea60b │ None              │
└───────────────────────┴───────────────┴──────────────────────────────────────┴───────────────────┘
```
</div>

> When you click into the `my-process-pool` you can click into the tab for work queues. You should see a red status icon listed for the default work queue signifying that this queue is not ready to submit work. Work queues are an advanced topic to help determine flow priority. You can learn more about work queues in the [work queue documentation.](https://docs.prefect.io/2.10.13/concepts/work-pools/#work-queues) 

Reviewer: ```suggestion
When you click into the `my-process-pool` you can click into the tab for work queues. You should see a red status icon listed for the default work queue signifying that this queue is not ready to submit work. Work queues are an advanced feature. You can learn more about them in the [work queue documentation.](https://docs.prefect.io/2.10.13/concepts/work-pools/#work-queues) 
```

To get the work queue healthy and ready to submit flow runs, you need to start a worker.

> Workers are a lightweight polling system that kick-off flow runs submitted to them by their work pool. To start your worker on your laptop you will open a new terminal and confirm that your virtual environment is activated. Run the following command in this new terminal to start the worker:

Reviewer: ```suggestion
Workers are a lightweight polling system that kick-off flow runs submitted to them by their work pool. To start a worker on your laptop, you will open a new terminal and confirm that your virtual environment is activated. Run the following command in this new terminal to start the worker:
```

> Workers are a lightweight polling system that kick-off flow runs submitted to them by their work pool. To start a worker on your laptop, you will open a new terminal and confirm that your virtual environment is activated. Run the following command in this new terminal to start the worker:

Reviewer: ```suggestion
Workers are a lightweight polling system that kick-off flow runs submitted to them by their work pool. To start a worker on your laptop, open a new terminal and confirm that your virtual environment is activated. Run the following command in this new terminal to start the worker:
```

> As mentioned above, workers are a lightweight polling system that kick-off flow runs submitted to them by their work pool. To start your worker on your laptop you will open a new terminal and double check that your virutual envionment is activated. Run the following command in this new terminal to start the worker:

Reviewer: ```suggestion
Workers are a lightweight polling system that kick-off flow runs submitted to them by their work pool. To start your worker on your laptop you will open a new terminal and confirm that your virtual environment is activated. Run the following command in this new terminal to start the worker:
```

<div class="terminal">
```bash
prefect worker start --pool my-process-pool
```
</div>

> You should see the worker start. It is now polling the Prefect API to see if there are any scheduled flow runs to kick off. You’ll see your new worker listed in the UI under the worker tab of the Work Pool page with a recent last polled date. You should also be able to see a healthy status indicator in the default work queue under the work queue tab.

Reviewer: ```suggestion
You should see the worker start. It is now polling the Prefect API to request any scheduled flow runs for it to start. You’ll see your new worker listed in the UI under the worker tab of the Work Pool page with a recent last polled date. You should also be able to see a healthy status indicator in the default work queue under the work queue tab.
```

You will need to keep this terminal running in order to have the worker continue to pick up jobs. Since you are running this worker locally, the worker will terminate if you close the terminal. For a production setting, this worker should be running as a damonized or managed process. See next steps for more information on this.

> Now that we’ve set up your work pool and worker, they are ready to kick off deployed flow runs. Lets build a deployment that sends work to your `my-process-pool`.

Reviewer: ```suggestion
Now that we’ve set up your work pool and worker, they are ready to kick off deployed flow runs. Lets deploy your  tutorial flow to `my-process-pool`.
```

## Create a Deployment

> ### From our previous steps we now have:

Reviewer: ```suggestion
### From our previous steps, we now have:
```

1. Name your deployment `my-deployment`
2. Type `n` for now, you can set up a schedule later
3. Select the work pool you just created, `my-process-pool`
4. When asked if you would like your workers to pull your flow code from its remote repository, select yes if you’ve been following along and defining your flow code script from within a GitHub repository:
    - __`y`__: Recommended: Prefect will automatically register your GitHub repo as the the remote location of this flow’s code. This means a worker started on any machine (for example: on your laptop, on your team-mate’s laptop, or in a cloud VM) will be able to facilitate execution of this deployed flow.
    - __`n`__: If you would like to continue this tutorial without the use of GitHub, thats ok, Prefect will always first look to see if the flow code exists locally before referring to remote flow code storage, so your local `my-process-pool` should have all it needs to complete the execution of this deployed flow.

!!! tip "Did you know?"
    A Prefect flow can have more than one deployment. This can be useful if you want your flow to run in different execution environments or have multiple different schedules. 

> As you continue to use Prefect, you'll likely author many different flows and deployments of them. Check out the next section to learn about defining deployment objects in a `deployment.yaml` file.

Reviewer: ```suggestion
As you continue to use Prefect, you'll likely author many different flows and deployments of them. Check out the next section to learn about defining deployments in a `deployment.yaml` file.
```

> As you continue to use Prefect you'll likely author many different flows and want to define a variety of deployments for them. Check out the next section to learn about defining deployment objects in a `deployment.yaml` file.

Reviewer: ```suggestion
As you continue to use Prefect, you'll likely author many different flows and deployments of them. Check out the next section to learn about defining deployment objects in a `deployment.yaml` file.
```
